client/components/theme-type-badge/tooltip.tsx

```diff
@@ -0,0 +1,177 @@
+import { recordTracksEvent } from '@automattic/calypso-analytics';
+import {
+	DOT_ORG_THEME,
+	MARKETPLACE_THEME,
+	PREMIUM_THEME,
+	WOOCOMMERCE_THEME,
+} from '@automattic/design-picker';
+import { Button as LinkButton } from '@wordpress/components';
+import { createInterpolateElement } from '@wordpress/element';
+import { useTranslate } from 'i18n-calypso';
+import { useEffect } from 'react';
+import { useSelector } from 'calypso/state';
+import {
+	canUseTheme,
+	getThemeType,
+	isThemePurchased,
+	isMarketplaceThemeSubscribed,
+	getMarketplaceThemeSubscriptionPrices,
+} from 'calypso/state/themes/selectors';
+import getSelectedSiteId from 'calypso/state/ui/selectors/get-selected-site-id';
+import getSelectedSiteSlug from 'calypso/state/ui/selectors/get-selected-site-slug';
+
+interface Props {
+	themeId: string;
+}
+
+const ThemeTypeBadgeTooltip = ( { themeId }: Props ) => {
+	const translate = useTranslate();
+	const siteId = useSelector( getSelectedSiteId );
+	const siteSlug = useSelector( getSelectedSiteSlug );
+	const type = useSelector( ( state ) => getThemeType( state, themeId ) );
+	const isIncludedCurrentPlan = useSelector(
+		( state ) => siteId && canUseTheme( state, siteId, themeId )
+	);
+	const isPurchased = useSelector( ( state ) => {
+		if ( ! siteId ) {
+			return false;
+		}
+
+		if ( isThemePurchased( state, themeId, siteId ) ) {
+			return true;
+		}
+
+		if ( type === MARKETPLACE_THEME ) {
+			return isMarketplaceThemeSubscribed( state, themeId, siteId );
+		}
+
+		return false;
+	} );
+	const subscriptionPrices = useSelector( ( state ) =>
+		type === MARKETPLACE_THEME ? getMarketplaceThemeSubscriptionPrices( state, themeId ) : {}
+	);
+
+	useEffect( () => {
+		recordTracksEvent( 'calypso_upgrade_nudge_impression', {
+			cta_name: 'theme-upsell-popup',
+			theme: themeId,
+		} );
+	}, [ themeId ] );
+
+	const goToCheckout = ( plan: string ) => {
+		recordTracksEvent( 'calypso_theme_tooltip_upgrade_nudge_click', { plan } );
+
+		if ( siteSlug ) {
+			const params = new URLSearchParams();
+			params.append( 'redirect_to', window.location.href.replace( window.location.origin, '' ) );
+
+			window.location.href = `/checkout/${ encodeURIComponent(
+				siteSlug
+			) }/${ plan }?${ params.toString() }`;
+		}
+	};
+
+	let header;
+	if ( type === DOT_ORG_THEME ) {
+		header = translate( 'Community theme', {
+			context: 'This theme is developed and supported by a community',
+			textOnly: true,
+		} );
+	} else if ( type === MARKETPLACE_THEME ) {
+		header = translate( 'Paid theme' );
+	} else if ( type === WOOCOMMERCE_THEME ) {
+		header = translate( 'WooCommerce theme' );
+	} else {
+		header = translate( 'Premium theme' );
+	}
```

Comment:
There is a lot of branching logic when determining the header and message. Maybe we could use a lookup map for retrieving the translations. This example has an algorithmic complexity of O(n) and has fewer conditional branches:

We would declare this outside of the component (since it doesn't share responsibility with it)
```
const getHeader = (type, translate) => {
  const headerMap = {
    [DOT_ORG_THEME]: translate('Community theme', {
      context: 'This theme is developed and supported by a community',
      textOnly: true,
    }),
    [MARKETPLACE_THEME]: translate('Paid theme'),
    [WOOCOMMERCE_THEME]: translate('WooCommerce theme'),
    [PREMIUM_THEME]: translate('Premium theme'),
  };

  return headerMap[type];
};
```

And then we could just do:
```
let header = getHeader( type, translate );
```

Which seems more readable and easy to maintain.

We could also apply this to the message and obtain the same benefits.

Reply:
Nice suggestion! Done.

> We could also apply this to the message and obtain the same benefits.

Not sure we can do that, since messages also depend on other variables such as `isPurchased` and `isIncludedCurrentPlan` which are hard to add in a lookup map.